File: src/main/java/org/myfintech/payment/service/impl/ContractServiceImpl.java
package org.myfintech.payment.service.impl;

import java.util.List;
import java.util.Optional;
import java.util.Set;

import org.myfintech.payment.domain.ContractCreateDTO;
import org.myfintech.payment.domain.ContractDTO;
import org.myfintech.payment.entity.Contract;
import org.myfintech.payment.entity.projection.ContractWithClientProjection;
import org.myfintech.payment.exception.Http404NotFoundException;
import org.myfintech.payment.mapper.ContractMapper;
import org.myfintech.payment.repository.ContractRepository;
import org.myfintech.payment.service.ClientService;
import org.myfintech.payment.service.ContractService;
import org.myfintech.payment.validator.CotractValidator;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import static org.myfintech.payment.exception.handler.ExceptionHandlerConstants.CONTRACT_NOT_FOUND;

@Service
public class ContractServiceImpl implements ContractService {

	private final ContractRepository repository;
	private final ContractMapper mapper;
	private final ClientService clientService;
	private final CotractValidator validator;

	public ContractServiceImpl(ContractRepository repository, ContractMapper mapper, ClientService clientService,
			CotractValidator validator) {
		this.repository = repository;
		this.mapper = mapper;
		this.clientService = clientService;
		this.validator = validator;
	}

	@Override
	public List<ContractDTO> findAll() {
		return repository.findAll().stream().map(mapper::toDTO).toList();
	}

	@Override
	@Transactional(readOnly = true)
	public ContractDTO findById(Long id) {
		return repository.findById(id).map(mapper::toDTO).orElseThrow();
	}

	@Override
	@Transactional
	public ContractDTO save(ContractCreateDTO dto) {
		return mapper.toDTO(repository.save(mapper.toEntity(dto, clientService.findEntityById(dto.clientId()))));
	}

	@Override
	@Transactional
	public ContractDTO update(Long id, ContractDTO dto) {
		Contract contract = repository.findById(id).orElseThrow(() -> new Http404NotFoundException(CONTRACT_NOT_FOUND + id));
		contract.setContractNumber(dto.contractNumber());
		return mapper.toDTO(contract);
	}

	@Override
	@Transactional(readOnly = true)
	public Optional<Contract> findByContractNumber(String contractNumber) {
		return Optional.of(repository.findByContractNumber(contractNumber)
				.orElseThrow(() -> new Http404NotFoundException(CONTRACT_NOT_FOUND + contractNumber)));
	}

	@Override
	@Transactional(readOnly = true)
	public List<ContractWithClientProjection> findAllByContractNumbers(Set<String> contractNumbers) {
		return repository.findAllByContractNumberIn(contractNumbers);
	}

	@Override
	@Transactional(propagation = Propagation.SUPPORTS)
	public void validateUpdateRequest(Long id, ContractDTO contractDTO) {
		validator.validateUpdateRequest(id, contractDTO);
		
	}

	@Override
	@Transactional(propagation = Propagation.SUPPORTS)
	public void validateCreate(ContractCreateDTO contractDTO) {
		validator.validateCreate(contractDTO);
		
	}
}

File: src/main/kotlin/org/myfintech/payment/service/impl/ContractServiceImpl.kt
package org.myfintech.payment.service.impl

import org.myfintech.payment.domain.*
import org.myfintech.payment.entity.Contract
import org.myfintech.payment.entity.projection.ContractWithClientProjection
import org.myfintech.payment.exception.Http404NotFoundException
import org.myfintech.payment.repository.ContractRepository
import org.myfintech.payment.service.ClientService
import org.myfintech.payment.service.ContractService
import org.myfintech.payment.validator.CotractValidator
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ContractServiceImpl(
    private val repository: ContractRepository,
    private val clientService: ClientService,
    private val validator: CotractValidator
) : ContractService {

    override fun findAll(): List<ContractDTO> {
        return repository.findAll().map { it.toDTO() }
    }

    @Transactional(readOnly = true)
    override fun findById(id: Long): ContractDTO {
        return repository.findByIdOrNull(id)?.toDTO()
            ?: throw Http404NotFoundException("Contract not found: $id")
    }

    @Transactional
    override fun save(dto: ContractCreateDTO): ContractDTO {
        val client = clientService.findEntityById(dto.clientId)
        val savedEntity = repository.save(dto.toEntity(client))
        return savedEntity.toDTO()
    }

    @Transactional
    override fun update(id: Long, dto: ContractDTO): ContractDTO {
        val contract = repository.findByIdOrNull(id)
            ?: throw Http404NotFoundException("Contract not found: $id")
        contract.contractNumber = dto.contractNumber
        return contract.toDTO()
    }

    @Transactional(readOnly = true)
    override fun findByContractNumber(contractNumber: String): Contract {
        return repository.findByContractNumber(contractNumber)
    }

    @Transactional(readOnly = true)
    override fun findAllByContractNumbers(contractNumbers: Set<String>): List<ContractWithClientProjection> {
        return repository.findAllByContractNumberIn(contractNumbers)
    }

    @Transactional(readOnly = true)
    override fun validateUpdateRequest(id: Long, contractDTO: ContractDTO) {
        validator.validateUpdateRequest(id, contractDTO)
    }

    @Transactional(readOnly = true)
    override fun validateCreate(contractDTO: ContractCreateDTO) {
        validator.validateCreate(contractDTO)
    }
}